### PR TITLE
Add Dependency Tracking

### PR DESCRIPTION
**Pull request by Mr4k**
Okay after a while this one should be it. 
Less confusing than the first attempt.
A little faster but still pretty slow (should be able to fix that in later versions).
Completely up to date with all language features.
Some amount of test coverage (needs more eventually).

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -2,8 +2,10 @@ bench:
 	cd benchmark && go test -bench=. -benchtime=30s
 fmt:
 	go fmt ./...
+profile:
+	cd benchmark && go test -cpuprofile cpu.prof -memprofile mem.prof -bench=. -benchtime=30s && go tool pprof -svg cpu.prof
 test:
-	go test ./...
+	go test -count=1 ./...
```
Comment (jemmaissroff) [inline Makefile:9]: Can we add a comment specifying that this disables caching?

File: benchmark/benchmark_test.go [hunk 2]
```diff
@@ -25,76 +28,74 @@ func BenchmarkFib(b *testing.B) {
...
+			strInput += ","
+		}
+	}
+	strInput += "]"
+	program, env := testBuild(fmt.Sprintf(`
 	let get_n_elements = fn(arr, offset, number_of_elements) { if (number_of_elements == 0) { [] } else { [arr[offset]] + get_n_elements(arr, offset + 1, number_of_elements - 1) } }
-	let merge_elements = fn(res_lower, res_upper) { if (len(res_lower) == 0) { if (len(res_upper) == 0) { [] } else { res_upper } } else { if (len(res_upper) == 0) { res_lower } else { if (first(res_upper) < first(res_lower)) { [first(res_upper)] + merge_elements(res_lower, rest(res_upper)) } else { [first(res_lower)] + merge_elements(res_upper, rest(res_lower)) } } } }
 
-	let merge_sort = fn(arr) { if (len(arr) == 1) { return arr } else { let half = int(len(arr)/2); let res_lower = get_n_elements(arr, 0, half); let res_upper = get_n_elements(arr, half, len(arr) - half); merge_elements(merge_sort(res_lower), merge_sort(res_upper)) } }
+	let car = fn(a) { a[0] }
```
Comment (jemmaissroff) [inline benchmark/benchmark_test.go]: Can we rename these methods to keep the naming consistent with stdlib? So `first`, `last`

File: object/object.go [hunk 3]
```diff
@@ -64,9 +102,17 @@ func (i *Integer) Equal(o Object) bool {
 	return ok && comp.Value == i.Value
 }
 func (i *Integer) Falsey() Object { return ZERO_INTEGER }
+func (i *Integer) AddDependency(dep Object) {
+	if dep == nil {
```
Comment (jemmaissroff) [inline object/object.go]: Remove?

File: object/object.go [hunk 4]
```diff
@@ -249,12 +361,30 @@ func (a *Array) Equal(o Object) bool {
...
+	a.Dependencies = append(a.Dependencies, dep)
+}
+func (a *Array) AddLengthDependency(dep Object) {
+	a.Length.AddDependency(dep)
+}
+func (a *Array) AddOffsetDependency(dep Object) {
+	a.Offset.AddDependency(dep)
+}
+
+func (a *Array) GetDependencyLinks() []Object {
```
Comment (jemmaissroff) [inline object/object.go:376]: Does an `Array` depend on its offset?

File: object/object.go [hunk 5]
```diff
@@ -312,7 +449,19 @@ type HashPair struct {
 }
 
 type Hash struct {
-	Pairs map[HashKey]HashPair
+	Pairs        map[HashKey]HashPair
```
Comment (jemmaissroff) [inline object/object.go:450]: Do hashes need an `Offset`?

File: object/object.go [hunk 6]
```diff
@@ -345,6 +494,14 @@ func (h *Hash) Equal(o Object) bool {
 }
 func (h *Hash) Falsey() Object { return EMPTY_HASH }
 
+func (h *Hash) Copy() Object {
+	return &Hash{Pairs: h.Pairs, Length: *h.Length.Copy().(*Integer), Dependencies: []Object{h}}
```
Comment (jemmaissroff) [inline object/object.go]: Can we make `Length` depend on previous `Length` instead of copying?

File: object/object.go [hunk 7]
```diff
@@ -249,12 +361,30 @@ func (a *Array) Equal(o Object) bool {
 	return true
 }
 func (a *Array) Falsey() Object { return EMPTY_ARRAY }
+func (a *Array) Copy() Object {
+	return &Array{Elements: a.Elements, Dependencies: []Object{a}, Length: *a.Length.Copy().(*Integer), Offset: *a.Offset.Copy().(*Integer)}
```
Comment (jemmaissroff) [inline object/object.go:363]: Can `Length` depend on parent's `Length`?

File: object/object.go [hunk 8]
```diff
@@ -353,3 +510,27 @@ type HashKey struct {
...
 }
+
+type DebugTraceMetadata struct {
+	DebugMetadata map[string]bool
+	Dependencies  []Object
+}
+
+// NOTE this object is techincal debt
+// TODO (Peter) remove this gracefully and replace with a better version
+func (d *DebugTraceMetadata) Type() ObjectType { return "DEBUGTRACEMETADATA" }
```
Comment (jemmaissroff) [inline object/object.go]: Can we pull this into the `const` up top?

File: evaluator/evaluator_dependency_tracing_test.go [hunk 9]
```diff
@@ -0,0 +1,288 @@
...
+}
+
+func strEquals(t *testing.T, testStr string, truthStr string) {
+	if testStr != truthStr {
+		t.Errorf("string has wrong value. got %s, want %s",
+			testStr, truthStr)
+	}
+}
+
+/*func assertProgramDidNotError(t *testing.T, res object.Object) object.Object {
```
Comment (jemmaissroff) [inline evaluator/evaluator_dependency_tracing_test.go]: Remove?

**Review by jemmaissroff — COMMENTED** [review]
Looking great!
Can we remove the `monkey` binary?

File: object/object.go (hunk 3 above)
Comment (Mr4k) [reply object/object.go]: lol

File: object/object.go (hunk 5 above)
Comment (Mr4k) [reply object/object.go:450]: yessssss?

File: object/object.go (hunk 4 above)
Comment (Mr4k) [reply object/object.go:376]: No I think

File: object/object.go (hunk 5 above)
Comment (Mr4k) [reply object/object.go:450]: confirmed yes

File: object/object.go (hunk 6 above)
Comment (Mr4k) [reply object/object.go]: This actually does that already given the copy semantics.

File: object/object.go (hunk 7 above)
Comment (Mr4k) [reply object/object.go:363]: No I think

File: object/object.go (hunk 8 above)
Comment (Mr4k) [reply object/object.go]: Done!

File: evaluator/evaluator_dependency_tracing_test.go (hunk 9 above)
Comment (Mr4k) [reply evaluator/evaluator_dependency_tracing_test.go]: 👍

Comment (Mr4k) [reply evaluator/evaluator_dependency_tracing_test.go]: goodbye dead code. You will be missed and lived a full life from mid November until now

**Comment by Mr4k** [comment]
woooooooooo

**Pull request merged**